File: exercism/dart/matching-brackets/lib/matching_brackets.dart
class MatchingBrackets {
  bool isPaired(String s) {
    int length = s.length;
    List<String> stack = List();
    for (int i = 0; i < length; i++) {
      switch (s[i]) {
        case "[":
          stack.add("[");
          break;
        case "{":
          stack.add("{");
          break;
        case "(":
          stack.add("(");
          break;
        case "]":
          if (stack.isEmpty) return false;
          String last = stack.removeLast();
          if (last != "[") return false;
          break;
        case "}":
          if (stack.isEmpty) return false;
          String last = stack.removeLast();
          if (last != "{") return false;
          break;
        case ")":
          if (stack.isEmpty) return false;
          String last = stack.removeLast();
          if (last != "(") return false;
          break;
        default:
          break;
      }
    }
    return stack.isEmpty;
  }
}

File: exercism/python/matching-brackets/matching_brackets.py
def is_paired(input_string):

    if len(input_string) == 0:
        return True

    pairing_char = {'}': '{', ')': '(', ']': '['}
    stack = []

    for c in input_string:
        if c in pairing_char.values():
            stack.append(c)
        elif c in pairing_char.keys() and (len(stack) == 0 or stack.pop() != pairing_char[c]):
            return False

    return len(stack) == 0
